### PR TITLE
Lineage version V2 with right APIs

### PR DESCRIPTION
**Pull request by nitinmotgi**

### CODE REVIEW

**Comment by nitinmotgi** [comment]
Will squash and merge once approved.

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Lineage.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+package io.cdap.wrangler.api.lineage;
+
+import io.cdap.cdap.api.annotation.Beta;
+
+/**
+ * This class <code>Lineage</code> defines a mutation defined by the directive that gets capture as lineage.
+ *
+ * <p>Directives have to implement this class to inject their mutations for lineage to be constructed.</p>
+ * <p>
+ *   The method <code>lineage</code> is invoked separately in the <code>preRun</code> phase of the pipeline
```
Comment (yaojiefeng) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Lineage.java]: nit: prepareRun

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2017-2019 Cask Data, Inc.
+ * Copyright © 2019 Cask Data, Inc.
```
Comment (yaojiefeng) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java]: This should still be 2017-2019 since the file is not added in this pr

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Relation.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+package io.cdap.wrangler.api.lineage;
+
+import io.cdap.cdap.api.annotation.Beta;
+
+import java.io.Serializable;
+import java.util.List;
+
+/**
+ * The <code>Relation</code> class represents a relation between a list of sources to a list of targets.
+ * The relation can have 4 different types and can be many-to-many, many-to-one, or one-to-many.
```
Comment (yaojiefeng) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Relation.java]: Missing one-to-one here.

File: wrangler-core/src/main/java/io/cdap/directives/column/Rename.java [hunk 4]
```diff
@@ -94,4 +85,12 @@ public void destroy() {
     }
     return rows;
   }
+
+  @Override
+  public Mutation lineage() {
+    return Mutation.builder()
+      .readable("Renamed column '%s' to '%s'", source.value(), target.value())
+      .relation(source, Many.columns(source, target))
```
Comment (yaojiefeng) [inline wrangler-core/src/main/java/io/cdap/directives/column/Rename.java]: this should be a one-to-one mapping since there is no source col after the rename

File: wrangler-core/src/main/java/io/cdap/directives/column/SplitToColumns.java [hunk 5]
```diff
@@ -99,5 +102,26 @@ public void destroy() {
...
+  @Override
+  public Mutation lineage() {
+    return Mutation.builder()
+      .readable("Split the column '%s' with regex '%s'", column, regex)
+      .relation(
+        column,
+        Many.columns(
+          column,
+          String.format("%s_%d", column, 1),
+          String.format("%s_%d", column, 2),
```
Comment (yaojiefeng) [inline wrangler-core/src/main/java/io/cdap/directives/column/SplitToColumns.java:115]: We are hard coding the columns here, should we use ALL relation in this case?

File: wrangler-core/src/main/java/io/cdap/wrangler/lineage/LineageOperations.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+   * Generates the list of {@link FieldOperation} required for generating lineage.
+   *
+   * @return a {@link List} of {@link FieldOperation}
+   *
+   * @throws Exception thrown if there is any issue generating field operations.
+   */
+  public List<FieldOperation> generate() throws Exception {
+    // After input and output schema are validated, it's time to extract
+    // all the fields from input and output schema.
+    Set<String> input = inputSchema.getFields().stream()
```
Comment (yaojiefeng) [inline wrangler-core/src/main/java/io/cdap/wrangler/lineage/LineageOperations.java]: We can change the constructor to accept Set<String> as the input and output fields instead of Schema, since in `Wrangler` we validate the input and output schema are not null, so it is better to have this logic there. Doing it here will have IDE warning about NPE

**Review by yaojiefeng — COMMENTED** [review]
Some minor comments

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Lineage.java (hunk 1 above)
Comment (nitinmotgi) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Lineage.java]: fixed

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java (hunk 2 above)
Comment (nitinmotgi) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java]: fixed

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Relation.java (hunk 3 above)
Comment (nitinmotgi) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Relation.java]: fixed

File: wrangler-core/src/main/java/io/cdap/directives/column/Rename.java (hunk 4 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/directives/column/Rename.java]: fixed

File: wrangler-core/src/main/java/io/cdap/directives/column/SplitToColumns.java (hunk 5 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/directives/column/SplitToColumns.java:115]: No, I would rather use fixed set. If there is nothing beyond this for a column, it would be ignored anyway.

File: wrangler-core/src/main/java/io/cdap/wrangler/lineage/LineageOperations.java (hunk 6 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/wrangler/lineage/LineageOperations.java]: fixed!

**Comment by nitinmotgi** [comment]
@yaojiefeng addressed all your comments.

**Review by yaojiefeng — APPROVED** [review]

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+ *    .readable("string")
+ *    .relation("source", Many.columns("col3", "col4", "col5").build();
+ * </code>
+ *
+ * @see Lineage
+ * @see Mutation
+ * @see Relation
+ */
+@Beta
+public final class Many implements Serializable {
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java:52]: should have serialVersionUID

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java [hunk 8]
```diff
@@ -16,60 +16,242 @@
...
+ * @see Many
  */
-public final class Mutation {
-  // Name of the column on which the mutation is applied.
-  private final String column;
-
-  // Specifies the type of the mutation being applied on the column.
-  private final MutationType type;
+@Beta
+public final class Mutation implements Serializable {
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java:43]: should have serialVersionUID

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Relation.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+ *     one-to-many.
+ *   </li>
+ * </ul>
+ *
+ * @see Lineage
+ * @see Many
+ * @see Mutation
+ */
+@Beta
+public final class Relation implements Serializable {
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Relation.java:53]: same here

File: wrangler-transform/src/main/java/io/cdap/wrangler/Wrangler.java [hunk 10]
```diff
@@ -517,31 +506,13 @@ private RecipeParser initializeAndGetParser(
...
 
-    // Make a call to retrieve the directive config and create a context.
-    // In this plugin, a call is made to the service at runtime.
-    // NOTE: This has to be moved to pre-start-plugin phase so that the information
-    // it's not heavily parallelized to crash the service.
-    try {
-      URL url = getDPServiceURL(context, CONFIG_METHOD);
-      if (url != null) {
-        ConfigDirectiveContext dContext = new ConfigDirectiveContext(url);
-        recipe.initialize(dContext);
```
Comment (CuriousVini) [inline wrangler-transform/src/main/java/io/cdap/wrangler/Wrangler.java:531]: why is this change needed?

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Lineage.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+ *   are available to be used within the <code>lineage</code> method.
+ * </p>
+ *
+ * <p>
+ *   {@link Mutation} captures all the changes the directive is going to be applying of the data. It has
+ *   two major methods:
+ *
+ *   <ul>
+ *     <il>
+ *       <code>readable</code> - This method is defined to provide the post readable description of the mutation
```
Comment (sreevatsanraman) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Lineage.java]: replace post readable with post transformation

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Lineage.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+ *     the directive is applying on data. Care should be taken to use the right tense as the lineage would be consumed
+ *     by users after the transformation has been applied on the data. As best practise, it's highly recommended to use
+ *     past-tense for describing the transformations. Additionally, the language of the description is not trying
+ *     to provide complete details and configuration, but actually focusing on operations that directive has
+ *     performed on data.
```
Comment (sreevatsanraman) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Lineage.java]: Can you add an example?

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    return new Many(columns);
+  }
+
+  /**
+   * List of {@link ColumnName} specified as source or target.
+   *
+   * @param columns list source or target columns of type {@link List} of {@link ColumnName}.
+   * @return a instance of {@link Many}.
+   */
+  public static Many columns(List<ColumnName> columns) {
```
Comment (sreevatsanraman) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java]: It is unclear when to use columns(List<ColumnName>) vs columns(List<String>). Can you provide guidance in the javadoc

Comment (nitinmotgi) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java]: One is List with ColumnName and other with String.

File: wrangler-transform/src/main/java/io/cdap/wrangler/Wrangler.java [hunk 14]
```diff
@@ -517,31 +506,13 @@ private RecipeParser initializeAndGetParser(
...
+      new MigrateToV2(directives).migrate(),
+      registry
+    );
 
-    // Make a call to retrieve the directive config and create a context.
-    // In this plugin, a call is made to the service at runtime.
-    // NOTE: This has to be moved to pre-start-plugin phase so that the information
-    // it's not heavily parallelized to crash the service.
-    try {
-      URL url = getDPServiceURL(context, CONFIG_METHOD);
```
Comment (CuriousVini) [inline wrangler-transform/src/main/java/io/cdap/wrangler/Wrangler.java:528]: also `getDPServiceURL ` is an unused method. It should be removed

File: wrangler-core/src/main/java/io/cdap/wrangler/lineage/LineageOperations.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+            definedSources.addAll(relation.getSources());
+          }
+        );
+      }
+    }
+
+    // We iterate through all the input fields in the schema, check if there is corresponding
+    // field in the output schema. If both exists, then a identity mapping transform is added
+    // to the {@code FieldTransformationOperation} is added.
+    input.removeAll(definedSources);
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/lineage/LineageOperations.java]: this will modify the input hashset that was passed to this class. we should make a copy of it in the constructor.

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Lineage.java [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+ * @see Many
+ */
+@Beta
+public interface Lineage {
+  /**
+   * Returns a Mutation that can be used to generate lineage.
+   *
+   * @return a instance of {@link Mutation} object used to generate lineage.
+   */
+  Mutation lineage();
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Lineage.java:86]: Would it be a good idea to call it `getMutation()`?

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+    return new Many(columns);
+  }
+
+  /**
+   * Set of columns as source or target.
+   *
+   * @param columns list source or target columns of type {@linkn Set} of {@link String}.
+   * @return a instance of {@link Many}.
+   */
+  public static Many of(Set<String> columns) {
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java]: we should use `Collection` instead

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+    return new Many(columns.stream().map(ColumnName::value).collect(Collectors.toList()));
+  }
+
+  /**
+   * Set of {@link ColumnName} specified as source or target.
+   *
+   * @param columns list source or target columns of type {@link Set} of {@link ColumnName}.
+   * @return a instance of {@link Many}.
+   */
+  public static Many columns(Set<ColumnName> columns) {
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java]: same here. We should use `Collection`

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+ * @see Relation
+ */
+@Beta
+public final class Many implements Serializable {
+  private List<String> columns = new ArrayList<>();
+
+  /**
+   * @return a instance of <pre>List<String></pre> that returns columns associated with source or target.
+   */
+  public List<String> columns() {
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java:59]: this method returns list of strings where as other methods return `Many`. Since this is a getter, how about calling this `getColumns` or `getColumnNames`.

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+   * @return a instance of {@link Many}.
+   */
+  public static Many of(List<String> columns) {
+    return new Many(columns);
+  }
+
+  /**
+   * Set of columns as source or target.
+   *
+   * @param columns list source or target columns of type {@linkn Set} of {@link String}.
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java]: typo: @link

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java [hunk 21]
```diff
@@ -16,60 +16,242 @@
...
+    this.relations = new ArrayList<>();
+  }
 
-  public Mutation(String column, MutationType type, String description) {
-    this.column = column;
-    this.type = type;
-    this.description = description;
+  private Mutation(String readable, List<Relation> relations) {
+    this.readable = readable;
+    this.relations = relations;
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java]: should do a deep copy of relations since the class is immutable.

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java [hunk 22]
```diff
@@ -16,60 +16,242 @@
...
   }
 
   /**
-   * @return the type of mutation being applied on the <code>column</code>.
+   * @return a {@link List} of {@link Relation} associated with the transformation.
    */
-  public MutationType type() {
-    return type;
+  public List<Relation> relations() {
+    return relations;
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java:68]: should return a deep copy of list of relations.

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Relation.java [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+  /**
+   * Specifies the type of relations that can exist between source and target.
+   */
+  public enum Type {
+    DROP,
+    ALL,
+    CREATE,
+    STANDARD,
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Relation.java:63]: why do we have `ALL` and `STANDARD` types?

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java [hunk 24]
```diff
@@ -16,60 +16,242 @@
...
+      return this;
+    }
+
+    /**
+     * Specifies a relation that has no cause but effect.
+     *
+     * @param targets a list of targets.
+     * @return a instance of {@link Mutation.Builder}.
+     */
+    public Mutation.Builder generate(Many targets) {
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java]: Is there any specific reason to use method name `generate`(equivalent of CREATE) and `relation` (equivalent of STANDARD). Would it be a good idea to use create and standard method names?

File: wrangler-core/src/main/java/io/cdap/wrangler/lineage/LineageOperations.java (hunk 15 above)
Comment (yaojiefeng) [reply wrangler-core/src/main/java/io/cdap/wrangler/lineage/LineageOperations.java]: Changed the implementation to not modify the hashset and make all the fields in the constructor immutable

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Lineage.java (hunk 16 above)
Comment (yaojiefeng) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Lineage.java:86]: Here since the class is named `Lineage`, I think it is good to call this `lineage`. Also, in wrangler, we don't have `get` as prefix for the method name. So Nitin made all the names in this pr not prefixed with `get`

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java (hunk 19 above)
Comment (yaojiefeng) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java:59]: In wrangler, we don't have `get` as prefix for the method name. So Nitin made all the names in this pr not prefixed with `get`

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java (hunk 20 above)
Comment (yaojiefeng) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Many.java]: I removed this method with the change for the parameter to Collection

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Relation.java (hunk 23 above)
Comment (yaojiefeng) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Relation.java:63]: STANDARD means the relation has to clear-defined, the source and target has to be specified. ALL means the directive will apply changes to all the output fields, which we don't know in the directive, so we will have special logic when we generate lineage for ALL operations

File: wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java (hunk 24 above)
Comment (yaojiefeng) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/lineage/Mutation.java]: I think `relation` is fine for STANDARD, since most of the directives will provide clear relation, will change `generate` to `create`

**Review by CuriousVini — APPROVED** [review]
